src/Microsoft.AspNetCore.Mvc.Core/Formatters/MediaType.cs

```diff
@@ -415,6 +425,37 @@ private static string CreateMediaTypeWithEncoding(StringSegment mediaType, Encod
             return $"{mediaType.Value}; charset={encoding.WebName}";
         }
 
+        private bool MatchesSubtypePattern(MediaType subtypePattern)
+        {
+            if (subtypePattern.MatchesAllSubTypes)
+            {
+                return true;
+            }
+
+            if (subtypePattern.SubTypeSuffix.HasValue)
+            {
+                if (SubTypeSuffix.HasValue)
+                {
+                    // Both the pattern and the media type being checked have suffixes, so both parts must match.
+                    var matchesSubtypeWithoutSuffix = subtypePattern.MatchesAllSubTypesWithoutSuffix ||
+                        subtypePattern.SubTypeWithoutSuffix.Equals(SubTypeWithoutSuffix, StringComparison.OrdinalIgnoreCase);
+                    var matchesSubtypeSuffix = subtypePattern.MatchesAllNonemptySubTypeSuffixes ||
+                        subtypePattern.SubTypeSuffix.Equals(SubTypeSuffix, StringComparison.OrdinalIgnoreCase);
+
+                    return matchesSubtypeWithoutSuffix && matchesSubtypeSuffix;
```

Comment:
Oh, I realise this logic could short-circuit the checks better (e.g., if `matchesSubtypeWithoutSuffix` is false, no need to evaluate `matchesSubtypeSuffix`). Will change it.